### ISSUE TITLE
"unknown error" adding customer

### ISSUE THREAD

**Issue by bsweitzer** [NCRCounterpointAPI/APIGuide#61]
REQUEST BODY:
{
  "AR_CUST": {
    "NAM": "Supermanddatetest4",
    "FST_NAM": "Clarkdatetest4",
    "LST_NAM": "Kendatestt",
            "EMAIL_ADRS_1": "ckent@supermandatetest947.com",
    
    "AR_SHIP_ADRS": [
      {
        "SHIP_ADRS_ID": "ADDRESS_1",
        "NAM": "HOME",
        "SHIP_NAM_TYP": "B"
      }
    ]
  }
}
 
RESPONSE:
{
    "ErrorCode": "ERROR_UNKNOWN",
    "Message": "An error occurred during AddCustomer: Object reference not set to an instance of an object.,    at System.Object.GetType()\r\n   at ServiceStack.OrmLite.OrmLiteReadCommandExtensions.SetFilter[T](IDbCommand dbCmd, String name, Object value)\r\n   at ServiceStack.OrmLite.OrmLiteReadCommandExtensions.SingleById[T](IDbCommand dbCmd, Object value)\r\n   at ServiceStack.OrmLite.OrmLiteExecFilter.Exec[T](IDbConnection dbConn, Func`2 filter)\r\n   at CPAPI.Repositories_85.CustomerRepository_85.CPAPI.Repositories_85.ICustomerRepository.GetCustomerById(String custNo)\r\n   at CPAPI.Business_85.CustomerBusiness_85.AddCustomer(AR_CUST_POST customerEdit, String workgroupId)"
}

Confirmed that the store has a default customer and workgroup is set up.  CounterPoint is version 8.5.2.1.

**Comment by MikeRitchie** [comment]
The JSON above worked in my test, so this may be config related. What workgroup did you check for?

Based on the stack trace in the response, I'd guess that the SY_WRKGRP record for your workgroup doesn't have a TMPLT_CUST_NO value. See #2 below.

There are 2 common areas for issues with the POST /Customer calls:
1. Getting the CUST_NO. If CUST_NO is not provided explicitly in the POSTed JSON, then the API will attempt to assign a CUST_NO based on standard Counterpoint logic. To do that, a workgroup is needed. A workgroup can be explicitly provided in the JSON as such, which I personally prefer if possible since it makes it very clear what workgroup is being used:
```
{
  "Workgroup": "1",
  "AR_CUST": {
    "NAM": "Superman",
    "FST_NAM": "Clark",
    "LST_NAM": "Kent",
    "STR_ID": "MAIN",
    "AR_CUST_NOTE": [
      {
        "Note_ID": "ADDED_85",
        "NOTE_TXT": "Be careful, Clark Kent is Superman."
      }
    ],
    "AR_SHIP_ADRS": [
      {
        "SHIP_ADRS_ID": "ADDRESS_1",
        "NAM": "HOME",
        "SHIP_NAM_TYP": "B"
      },
      {
        "SHIP_ADRS_ID": "ADDRESS_2",
        "NAM": "WORK",
        "SHIP_NAM_TYP": "B"
      }
    ]
  }
}
```
If the workgroup isn't provided explicitly, the API will get the workgroup for the Counterpoint user being used to authorize the request (the user passed in as part of the authorization header). Specifically, the SY_USR.WRKGRP_ID value is used.

Once a workgroup is known,  the CUST_NO is retrieved from SY_WRKGRP.NXT_CUST_NO and incremented.

2. The other thing that happens is default values are popultated in the customer for any values not provided explicitly in the JSON. These values come from the template customer. The template customer comes from SY_WRKGRP.TMPLT_CUST_NO for the given workgroup as described above. Note that once template customer data is combined with customer data from the JSON, that all required fields must be present. If there are still required fields that aren't provided either by the template customer record or the JSON, the insert will fail.

**Comment by Bflatt** [comment]
You must have a store id.  The workgroup your api counterpoint user is using must be assigned to a workgroup with a valid template customer and a next customer # for assignment.   2.0 api has bug on customer insert to all versions before 8.5.2.1

**Comment by Bflatt** [comment]
This is broken in versions before 8.5.2.1
You must upgrade to 8.5.2.1 or better.


[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by langillerr** [comment]
We're actually on 8.5.2.1 right now with DB version 31.12.

**Comment by Bflatt** [comment]
Make sure the api counterpoint user you are using is in a workgroup with a template customer and a next customer number.

[cid:image001.png@01D3DE08.34A85CB0]

[cid:image002.png@01D3DE08.34A85CB0]


[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by langillerr** [comment]
Sending our "ecommerce" workgroup which has the following assigned.

![image](https://user-images.githubusercontent.com/10121243/39364638-9fc41a24-49fc-11e8-9dbd-beab993c191d.png)

![image](https://user-images.githubusercontent.com/10121243/39364653-ae9f8588-49fc-11e8-8fd6-cf3128653e6f.png)

**Comment by Bflatt** [comment]
Could you share you input text for review?  This is my basic test script I use to test each of our sites after API install.

If inserting addresses be sure to not exceed standard Field lengths.  This will break the insert.   I had this problem with addresses entered by online customers initially.

{
  "AR_CUST": {
    "NAM": "Superman",
    "FST_NAM": "Clark",
    "LST_NAM": "Kent2",
    "STR_ID": "cpo",
    "AR_CUST_NOTE": [
      {
        "Note_ID": "ADDED_85",
        "NOTE_TXT": "Be careful, Clark Kent is Superman."
      }
    ],
    "AR_SHIP_ADRS": [
      {
        "SHIP_ADRS_ID": "ADDRESS_1",
        "NAM": "HOME",
        "SHIP_NAM_TYP": "B"
      },
      {
        "SHIP_ADRS_ID": "ADDRESS_2",
        "NAM": "WORK",
        "SHIP_NAM_TYP": "B"
      }
    ]
  }
}



[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by langillerr** [comment]
Sure, I'm currently using the one provided by the postman collection - though I have tried other addresses as well. I've also tried adding the workgroup via a header, and in the json.

{
	"WORKGROUP": "ECOMMERCE",
	"AR_CUST": {
		"CUST_TYP": "C",
		"NAM": "Happy Gilmore",
		"FST_NAM": "Happy",
		"LST_NAM": "Gilmore",
		"SALUTATION": "Mr.",
		"CUST_NAM_TYP": "P",
		"SLS_REP": "langillerr",
		"STR_ID": "E",
		"AR_CUST_NOTE": [{
			"NOTE": "{\\rtf1\\ansi\\deff0{\\fonttbl{\\f0\\fnil\\fcharset0 Arial;}}\n\\viewkind4\\uc1\\pard\\lang1033\\fs16 Have Bill verify his current phone number.  (we could not reach him last week)\\par\n\\par\nAH\\par\n}"
		}],
		"AR_SHIP_ADRS": [{
			"NAM": "Home address",
			"SHIP_ADRS_ID": "HOME",
			"SHIP_NAM_TYP": "P",
			"FST_NAM": "Grandma",
			"LST_NAM": "Gilmore",
			"SALUTATION": "Mrs.",
			"ADRS_1": "1 Waterford Way",
			"CITY": "Topeka",
			"STATE": "KA",
			"ZIP_COD": 10101
		}]
	}
}`

**Comment by Bflatt** [comment]
Leave out the workgroup.  It is not required.  I have never sent this.
See if that helps.

[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by langillerr** [comment]
Also, I just tried using yours (changed str_id) and it still resulted in the same error

**Comment by langillerr** [comment]
And yours did not include the workgroup.

**Comment by Bflatt** [comment]
All transactions entered from the api are governed by the counterpoint user used in the api call’s workgroup and permissions.


[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by Bflatt** [comment]
You may have a permissions issues in the API admin.  Restart the api and also check the api roles that you have add customer rights.

[cid:image001.jpg@01D2DEDA.41EDDC70]

Bill Flatt
Mainspring, LLC
6060 S American Plaza St E | Suite 300 | Tulsa, OK | 74135
Phone: 800.865.1427 x109 | Local: 918.317.0707
bflatt@mainspring.net<mailto:bflatt@mainspring.net> | www.mainspring.net<http://www.mainspring.net/> | Need Help?<mailto:support@mainspring.net>
[cid:image002.png@01D2DEDA.41EDDC70]<https://www.facebook.com/MainSpringRetailSolutions/>[cid:image003.png@01D2DEDA.41EDDC70]<https://twitter.com/MainspringPOS> [cid:image004.png@01D2DEDA.41EDDC70] <http://www.linkedin.com/company/576462>  [cid:image005.png@01D2DEDA.41EDDC70] <mailto:support@mainspring.net>

**Comment by langillerr** [comment]
Understood - I only tried using the workgroup hoping it would help here.  I don't actually send it anywhere else.

Here are the permissions I've setup for our "api" role.  The correct user is assigned to the role.

![image](https://user-images.githubusercontent.com/10121243/39365194-7a5147d8-49fe-11e8-8832-3f4bb94df67c.png)

**Comment by langillerr** [comment]
just for fun I turned them all on for that role, and no difference

**Comment by MikeRitchie** [comment]
This will be addressed in an upcoming 2.1 release coming very soon. Date will be shared as soon as it is available.
